File: pruefstuecke/anlage.mjs
/**
 * Prüfstück: Diese App gehört EINER Anlage – der von DariaTech.
 *
 * DER ANLASS ist eine Ansage des Inhabers vom 30.08.2026: „Beide Apps
 * müssen nur für DariaTech sein, also es sollen alle Serverdaten im
 * Hintergrund hinterlegt sein."
 *
 * VORGEFUNDEN war das Gegenteil: ein allgemeines Softphone. Zehn
 * Anbietervorlagen von easybell über sipgate bis zur FRITZ!Box, ein
 * freies Feld für die Serveradresse, eine Auswahl für den Transport.
 * Wer die App eines DariaTech-Kunden einrichtete, musste sich durch
 * eine Anbieterliste arbeiten, in der neun Einträge falsch waren – und
 * die zehn Zeilen Hinweistext zur richtigen Adresse waren nötig, weil
 * genau das reihenweise schiefging.
 *
 * WAS HIER FESTGENAGELT WIRD:
 *
 *  1. Die Serveradresse steht an GENAU EINER Stelle im Quelltext
 *     (`Anlage.kt`) und nirgends sonst. Ein Wert an zwei Stellen
 *     driftet an einer davon.
 *  2. Kein Eingabefeld mehr für Server, Transport oder Anbieter. Was
 *     der Nutzer nicht eintragen kann, kann er auch nicht falsch
 *     eintragen.
 *  3. Kein fremder Anbieter mehr im Quelltext.
 *  4. Ein Altbestand mit abweichendem Server wird GEMELDET, nicht
 *     stillschweigend übergangen. Wer seine App vor dieser Änderung auf
 *     einen anderen Server gestellt hatte, soll erfahren, warum sie
 *     jetzt woanders anruft.
 *
 * Aufruf:  node pruefstuecke/anlage.mjs
 */
import { readFileSync, existsSync, readdirSync } from "node:fs";

const ergebnisse = [];
const pruefe = (name, ok, extra = "") => {
  ergebnisse.push([name, ok]);
  console.log((ok ? "  ok   " : " FEHL  ") + name + (ok || !extra ? "" : "  – " + extra));
};

const QUELLEN = "app/src/main/java/de/dariatech/softphone";
const LAYOUT = "app/src/main/res/layout/activity_main.xml";
const dateien = existsSync(QUELLEN)
  ? readdirSync(QUELLEN).filter((d) => d.endsWith(".kt"))
  : [];
const lies = (d) => readFileSync(`${QUELLEN}/${d}`, "utf-8");

console.log("\n1) Es gibt eine Datei, in der die Anlage steht");
{
  pruefe("Anlage.kt liegt vor", dateien.includes("Anlage.kt"),
    "hier gehören Serveradresse und Transport hin – und nur hier");
  if (dateien.includes("Anlage.kt")) {
    const t = lies("Anlage.kt");
    /* DER SIP-NAME IST sip.dariatech.de – UND DAS IST KEIN GESCHMACK.
       Befund vom 05.09.2026, nachgemessen im Namensdienst:

         pbx.dariatech.de → 104.21.14.53, 172.67.157.219,
                            2606:4700:3030::6815:e35   (Cloudflare)
         sip.dariatech.de → 178.254.6.5                (die Anlage)

       Cloudflares Proxy führt HTTP und HTTPS weiter, sonst nichts – ein
       REGISTER über UDP kommt dort nie an. Wer diesen Namen einträgt,
       baut eine App, die sich niemals anmelden kann, und der Fehler
       sieht aus wie ein Netzproblem beim Kunden.

       DASSELBE STEHT IM iOS-REPO. Zwei Apps mit zwei Servern wären zwei
       Fehlerbilder für dieselbe Ursache. */
    const ohneKommentar = t.replace(/\/\*[\s\S]*?\*\//g, "").replace(/\/\/[^\n]*/g, "");
    pruefe("und sie nennt die SIP-Adresse der Anlage", /"sip\.dariatech\.de"/.test(t),
      (t.match(/const val SERVER = "([^"]*)"/) ?? [])[1] ?? "keine gefunden");
    pruefe("und NICHT den Namen hinter dem CDN",
      !/pbx\.dariatech\.de|portal\.dariatech\.de/.test(ohneKommentar),
      "Cloudflare führt nur HTTP/HTTPS weiter – SIP kommt dort nie an");
    pruefe("und den Transport", /"UDP"|"TCP"|"TLS"/.test(t));
  }
}

console.log("\n2) Die Serveradresse steht nur an dieser einen Stelle");
{
  const woanders = dateien
    .filter((d) => d !== "Anlage.kt")
    .filter((d) => /sip\.dariatech\.de/.test(lies(d)));
  pruefe("kein zweites Vorkommen im Kotlin-Quelltext",
    woanders.length === 0, woanders.join(", "));
}

console.log("\n3) Keine fremden Anbieter mehr");
{
  // Die Adressen aus der alten Vorlagenliste. Sie gehören in ein
  // allgemeines Softphone, nicht in die App EINES Anbieters.
  const fremd = [
    "easybell.de", "sipgate.de", "t-online.de", "fpbx.de",
    "fonial.de", "1und1.de", "fritz.box"
  ];
  const gefunden = [];
  for (const d of dateien) {
    const t = lies(d);
    for (const f of fremd) if (t.includes(f)) gefunden.push(`${d}: ${f}`);
  }
  pruefe("kein fremder Registrar im Quelltext", gefunden.length === 0, gefunden.join(" · "));
  pruefe("die Vorlagenliste ist weg", !dateien.includes("Presets.kt"),
    "Presets.kt liegt noch da");
}

console.log("\n4) Nichts davon lässt sich mehr eintippen");
{
  const layout = existsSync(LAYOUT) ? readFileSync(LAYOUT, "utf-8") : "";
  for (const [feld, name] of [
    ["@+id/domain", "kein Feld für die Serveradresse"],
    ["@+id/transport", "keine Auswahl für den Transport"],
    ["@+id/preset", "keine Anbieterliste"],
    ["@+id/presetHint", "kein Hinweistext zur Anbieterliste"]
  ]) {
    pruefe(name, !layout.includes(feld), feld);
  }
  // Was BLEIBEN muss: Ohne diese beiden Felder kann sich niemand mehr
  // anmelden. Ein Prüfstück, das nur Entfernen prüft, wird grün, wenn
  // jemand die halbe Ansicht löscht.
  pruefe("das Feld für den Benutzernamen bleibt", layout.includes("@+id/username"));
  pruefe("das Feld für das Passwort bleibt", layout.includes("@+id/password"));
  pruefe("der Knopf zum Verbinden bleibt", layout.includes("@+id/connectButton"));
}

console.log("\n5) Der Code liest den Server nicht mehr aus den Einstellungen");
{
  const gefunden = [];
  for (const d of dateien) {
    const t = lies(d);
    for (const m of t.matchAll(/(get|put)String\(\s*"(domain|transport|preset)"/g)) {
      gefunden.push(`${d}: ${m[0]}`);
    }
  }
  pruefe("weder gelesen noch geschrieben", gefunden.length === 0, gefunden.join(" · "));
}

console.log("\n6) Ein Altbestand wird gemeldet, nicht verschwiegen");
{
  // Wer die App vor dieser Änderung auf einen anderen Server gestellt
  // hatte, ruft ab jetzt woanders an. Das ist eine Nebenwirkung, und
  // eine Nebenwirkung gehört benannt – ins Protokoll, damit sie bei
  // einer Störungssuche auffindbar ist.
  const t = dateien.includes("Anlage.kt") ? lies("Anlage.kt") : "";
  pruefe("Anlage.kt sieht sich einen Altbestand an",
    /Altbestand/i.test(t) && /Log\.[wi]/.test(t),
    "ohne Meldung sucht jemand stundenlang, warum die App den Server gewechselt hat");
}

const fehler = ergebnisse.filter(([, ok]) => !ok);
console.log(`\n${ergebnisse.length - fehler.length}/${ergebnisse.length} Prüfungen bestanden`);
if (fehler.length > 0) console.log("FEHLGESCHLAGEN: " + fehler.map(([n]) => n).join(", "));
process.exit(fehler.length > 0 ? 1 : 0);

File: app/src/main/java/de/dariatech/softphone/Anlage.kt
package de.dariatech.softphone

import android.content.Context
import android.content.SharedPreferences
import android.util.Log

/**
 * Die Anlage, zu der diese App gehört – an EINER Stelle.
 *
 * DER ANLASS ist eine Ansage des Inhabers vom 30.08.2026: Die App ist
 * die App von DariaTech, nicht ein allgemeines Softphone. Alle
 * Serverdaten liegen im Hintergrund; der Nutzer trägt nur noch seinen
 * Benutzernamen und sein Passwort ein.
 *
 * WARUM DAS BESSER IST, nicht nur kürzer: Vorher stand hier eine Liste
 * von zehn Anbietervorlagen, ein freies Feld für die Serveradresse und
 * eine Auswahl für den Transport. Neun der zehn Vorlagen waren für
 * einen DariaTech-Kunden falsch, und der Hinweistext zur richtigen
 * Vorlage war zehn Zeilen lang – er war so lang, weil genau diese
 * Eingabe reihenweise schiefging (Portaladresse statt SIP-Adresse,
 * Cloudflare-Tunnel, Endpunktname statt SIP-Benutzername). Was man
 * nicht eintragen kann, kann man auch nicht falsch eintragen.
 *
 * WER HIER ETWAS ÄNDERT, ÄNDERT ES FÜR ALLE. Ein zweiter Ort mit
 * derselben Adresse driftet an einem von beiden; das Prüfstück
 * `anlage.mjs` lässt genau das scheitern.
 *
 * FÜR EINE EIGENE ANLAGE (Self-Hosting) ist das hier die einzige
 * Stelle, die ein eigener Bau anfassen muss – eine Zeile. Die App holt
 * sich die Adresse NICHT von irgendwoher; das wäre ein Weg, auf dem
 * jemand ein fremdes Ziel unterschieben könnte.
 */
object Anlage {
    /**
     * Der SIP-Server der DariaTech-Telefonanlage.
     *
     * ACHTUNG, DAS IST NICHT ZWANGSLÄUFIG DIE PORTALADRESSE: Wer das
     * Portal über einen Tunnel oder einen Zugangsschutz erreicht, gilt
     * das nur für die Weboberfläche – SIP läuft dort nicht durch.
     * Maßgeblich ist die Adresse, die das Portal bei den Nebenstellen
     * als SIP-Server anzeigt.
     *
     * UND GENAU DAS IST HIER PASSIERT. Bis zum 05.09.2026 stand hier
     * `pbx.dariatech.de` – der Name, unter dem die Anlage in den
     * Betriebsdokumenten geführt wird. Nachgemessen im Namensdienst:
     *
     *     pbx.dariatech.de → 104.21.14.53, 172.67.157.219,
     *                        2606:4700:3030::6815:e35
     *     sip.dariatech.de → 178.254.6.5
     *
     * Die erste Reihe gehört Cloudflare. Deren Proxy führt HTTP und
     * HTTPS weiter und sonst nichts – ein REGISTER über UDP kommt dort
     * nie an. Eine App mit diesem Namen kann sich nicht anmelden, und
     * der Fehler sieht für den Kunden aus wie ein Netzproblem bei ihm.
     *
     * Der SIP-Name ist `sip.dariatech.de`; so steht er auch im
     * Störungsprotokoll der Anlage (REGISTER von
     * <sip:…@sip.dariatech.de>). Wer die Anlage umzieht, ändert diesen
     * Namen im Namensdienst – nicht diese Zeile.
     *
     * DIESELBE ZEILE STEHT IN Anlage.swift der iOS-App. Wer hier
     * ändert, ändert dort mit.
     */
    /* UMGEZOGEN AM 08.09.2026 auf den neuen Hauptserver.
     *
     * DER VORFALL (Inhaber, 08.09.2026, mit Bildschirmfoto aus der
     * iOS-App – die Android-App trug denselben Namen): „Die App hat noch
     * keinen Ausweis von der Anlage." Kein eigenes Bild, keine Kontakte,
     * kein Chat, während die Telefonie längst über pbx.taamas.de lief.
     *
     * DER GRUND: Die App holt ihren Ausweis über
     * POST https://<server>:8443/token – und <server> war noch der ALTE
     * Rechner. Nachgemessen am selben Abend im Namensdienst:
     *
     *     sip.dariatech.de → 178.254.6.5    (alt, Asterisk dort AUS)
     *     pbx.taamas.de    → 31.70.137.58   (der neue Hauptserver)
     *
     * Der Inhaber hat den Namen am selben Tag festgelegt: „ich habe mich
     * für pbx.taamas.de entschieden für client für SIP anmeldung."
     * Altlasten vom alten Server werden ausdrücklich nicht übernommen.
     */
    const val SERVER = "pbx.taamas.de"

    /**
     * TLS – und nichts anderes. Seit dem 09.09.2026.
     *
     * DER AUFTRAG (Inhaber, nach einem Tag ohne Telefonie): „TLS muss
     * für die gesamte Anlage und alle Kunden verpflichtend sein, es
     * soll keine andere Wahl geben. Alles innerhalb der Anlage muss
     * zwangsläufig verschlüsselt sein."
     *
     * Bis dahin stand hier „TLS zuerst, nach acht Sekunden Rückfall auf
     * UDP" – mit einem Transportgedächtnis und einem Knopf „Trotzdem
     * unverschlüsselt anmelden". Die Anlage nimmt seit demselben Tag
     * jede Nebenstelle NUR noch über TLS an (transport=transport-tls
     * im Endpunkt). Ein Rückfall meldete sich damit ins Leere – die App
     * zeigte „abgemeldet", wo „TLS blockiert" die Wahrheit gewesen
     * wäre. Und er war vom Netz erzwingbar: Wer im selben WLAN Pakete
     * nach 5061 verwirft, hätte die App auf offenes UDP bekommen, samt
     * der SDES-Schlüssel im SDP.
     *
     * Kommt TLS nicht zustande, bleibt die App also unangemeldet und
     * sagt es. Das ist die Entscheidung des Inhabers – lieber sichtbar
     * nicht angemeldet als still offen.
     *
     * DIESELBE ZEILE STEHT IN Anlage.swift der iOS-App.
     */
    const val TRANSPORT = "TLS"

    /**
     * Der Dienst der Anlage für alles, was NICHT Telefonie ist:
     * Kollegen, Bilder, Kontakte, Faxe.
     *
     * EIGENER PORT, EIGENE VERSCHLÜSSELUNG. Er ist bewusst getrennt vom
     * Portal – das lauscht nur auf 127.0.0.1. Vorgabe ist 8443, und TLS
     * ist dort Pflicht: Ein Bearer-Token im Klartext über das
     * öffentliche Netz ist ein verschenktes Geheimnis, einmal
     * mitgelesen gilt es für immer (pbx/docs/CLIENT-API.md).
     *
     * DIESELBE ZEILE STEHT IN Anlage.swift der iOS-App.
     */
    const val DIENST = "https://$SERVER:8443"

    /** Was in den Einstellungen als Server dasteht – nur zum Ansehen. */
    val anzeige: String get() = "$SERVER ($TRANSPORT)"

    private const val ALT_SERVER = "domain"
    private const val ALT_TRANSPORT = "transport"
    private const val ALT_VORLAGE = "preset"

    /**
     * Einen Altbestand aufräumen – und ihn MELDEN, wenn er abwich.
     *
     * Wer die App vor dieser Änderung auf einen anderen Server gestellt
     * hatte, ruft ab jetzt woanders an. Das ist eine echte
     * Nebenwirkung, und eine Nebenwirkung wird nicht verschwiegen: Sie
     * steht im Protokoll, damit sie bei einer Störungssuche auffindbar
     * ist. Stillschweigend umzuschalten wäre die Sorte Fehler, die man
     * erst nach Stunden findet.
     *
     * Die alten Schlüssel werden danach entfernt – ein toter Wert in
     * den Einstellungen ist eine Falle für den Nächsten, der ihn liest
     * und für maßgeblich hält.
     */
    fun raeumeAltbestandAuf(prefs: SharedPreferences) {
        val alt = prefs.getString(ALT_SERVER, null)
        if (!alt.isNullOrBlank() && alt != SERVER) {
            Log.w(
                "Anlage",
                "Altbestand: In den Einstellungen stand der Server \"$alt\". " +
                    "Diese App ist auf $SERVER festgelegt und meldet sich ab jetzt " +
                    "dort an. Der alte Eintrag wird entfernt."
            )
        }
        if (prefs.contains(ALT_SERVER) || prefs.contains(ALT_TRANSPORT) ||
            prefs.contains(ALT_VORLAGE)
        ) {
            prefs.edit()
                .remove(ALT_SERVER)
                .remove(ALT_TRANSPORT)
                .remove(ALT_VORLAGE)
                .apply()
        }
    }
}

